tests/jobs/localisation/example_case/scripts/common_functions.py

```diff
@@ -0,0 +1,548 @@
+            "correlation_azimuth": 45.0,
+            "correlation_dip": 0.0,
+            "correlation_exponent": 1.9,
+            "trend_use": 0,
+            "trend_params": [1.0, -1.0],
+            "trend_relstd": 0.05,
+            "grid_dimension": [150, 250, 1],
+            "grid_file_name": "GRID.EGRID",
+        },
+        "response": {
+            "grid_dimension": [15, 25, 1],
+            "upscaled_file_name": "Upscaled.roff",
+            "grid_file_name": "UpscaleGrid.EGRID",
+            "response_function": "average",
+            "gen_data_file_name": "UpscaledField_0.txt",
+            "calculate_all_cells": True,
+        },
+        "observation": {
+            "directory": "observations",
+            "file_name": "observations.obs",
+            "data_dir": "obs_data",
+            "3D_param_file_name": "init_files/UpscaledObsField.roff",
+            "rel_error": 0.10,
+            "min_abs_error": 0.01,
+            "selected_grid_cells": [
+                [5, 10, 1],
+                [10, 20, 1],
+            ],
+            # "selected_grid_cells":[
+            #         [15,  1, 1],
+            #         [15, 25, 1],
+            #         [ 1, 25, 1],
+            #         [ 1,  1, 1],
+            #         [ 3,  3, 1],
+            #         [ 5, 23, 1],
+            #         [11,  4, 1],
+            #         [ 3, 12, 1],
+            #         [13, 18, 1],
+            #     ],
+        },
+        "optional": {
+            "write_upscaled_to_file": True,
+            "write_obs_pred_diff_field_file": True,
+        },
+    }
+
+    settings = copy.deepcopy(default_settings)
+    if spec_dict is not None:
+        main_key = "settings"
+        if main_key not in spec_dict:
+            raise KeyError(f"Missing main keyword {main_key} in file: {yml_file_name}")
+        settings_yml_dict = spec_dict[main_key]
+        valid_keys1 = default_settings.keys()
+        for key1 in settings_yml_dict.keys():
+            if key1 in valid_keys1:
+                valid_keys2 = default_settings[key1].keys()
+                for key2 in settings_yml_dict[key1].keys():
+                    if key2 in valid_keys2:
+                        print(
+                            f"Modifying default setting for keyword '{key2}' "
+                            f"under '{key1}' in file {yml_file_name}  "
+                        )
+                        settings[key1][key2] = settings_yml_dict[key1][key2]
+                    else:
+                        raise KeyError(
+                            f"Unknown keyword '{key2}' specified under keyword '{key1}'"
+                            f" in {yml_file_name}"
+                        )
+            else:
+                raise KeyError(
+                    f"Unknown keyword '{key1}' specified under keyword '{main_key}'"
+                    f" in {yml_file_name}"
+                )
+    return settings
+
+
+def read_test_config(config_file_name):
```

Comment:
I think this can be improved somewhat by 1. early return and 2. using try and catch to get rid of the if-statement.
Something like this:

```python
import yaml

def read_test_config(config_file_name):
    if config_file_name is None:
        return specify_settings(None, None)

    try:
        with open(config_file_name, "r", encoding="utf-8") as yml_file:
            spec_dict = yaml.safe_load(yml_file)
    except FileNotFoundError:
        raise IOError(f"File {config_file_name} not found.")
    except yaml.YAMLError as e:
        raise ValueError(f"Failed to parse YAML file {config_file_name}. Ensure it's a valid .yml or .yaml file. Error: {e}")

    return specify_settings(spec_dict, config_file_name)

```

Reply:
This function is removed in the updated version. Not used.